### PR TITLE
Check Wallet DB integrity

### PR DESCRIPTION
**Pull request by aqk**
Introduce new command `chia wallet check`

Use example:

```
chia wallet check --db-path blockchain_wallet_v2_r1_mainnet_3314004000.OLD.sqlite
Reading blockchain_wallet_v2_r1_mainnet_3314004000.OLD.sqlite

    ---- Errors Found ----
Missing Wallet IDs: [2, 4, 6]
Wallet IDs with no derivations: [5, 7]
unhardened Wallet ID 15 uses type 8 in derivation_paths, but type 6 in wallet table at these derivation indices: [497, 498]
  hardened Wallet ID 15 uses type 8 in derivation_paths, but type 6 in wallet table at these derivation indices: [497, 498]
unhardened Wallet ID 16 uses type 8 in derivation_paths, but type 6 in wallet table at these derivation indices: [508, 509, 510]
  hardened Wallet ID 16 uses type 8 in derivation_paths, but type 6 in wallet table at these derivation indices: [508, 509, 510]
Missing DerivationPaths for   hardened wallet_id=1: [496, 507]
Missing DerivationPaths for unhardened wallet_id=1: [496, 507]
Missing DerivationPaths for   hardened wallet_id=3: [496, 507]
Missing DerivationPaths for unhardened wallet_id=3: [496, 507]
Missing DerivationPaths for   hardened wallet_id=8: [496, 507]
Missing DerivationPaths for unhardened wallet_id=8: [496, 507]
Missing DerivationPaths for   hardened wallet_id=9: [496, 507]
Missing DerivationPaths for unhardened wallet_id=9: [496, 507]
Missing DerivationPaths for   hardened wallet_id=10: [496, 507]
Missing DerivationPaths for unhardened wallet_id=10: [496, 507]
Missing DerivationPaths for   hardened wallet_id=11: [496, 507]
Missing DerivationPaths for unhardened wallet_id=11: [496, 507]
Missing DerivationPaths for   hardened wallet_id=12: [496, 507]
Missing DerivationPaths for unhardened wallet_id=12: [496, 507]
Missing DerivationPaths for   hardened wallet_id=13: [496, 507]
Missing DerivationPaths for unhardened wallet_id=13: [496, 507]
Missing DerivationPaths for   hardened wallet_id=14: [496, 507]
Missing DerivationPaths for unhardened wallet_id=14: [496, 507]
Missing DerivationPaths for   hardened wallet_id=15: [0 to 496, 507]
Missing DerivationPaths for unhardened wallet_id=15: [0 to 496, 507]
Missing DerivationPaths for   hardened wallet_id=16: [0 to 507]
Missing DerivationPaths for unhardened wallet_id=16: [0 to 507]
```

### CODE REVIEW

**Comment by aqk** [comment]
This is still a bit WIP - I need to move the tests around

File: chia/cmds/check_wallet_db.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+
+if __name__ == "__main__":
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(scan("", sys.argv[1]))
+
+
+def test_check_for_gaps_end_lt_start() -> None:
+    with pytest.raises(ValueError):
+        _ = check_for_gaps([], 2, 0)
```
Comment (arvidn) [inline chia/cmds/check_wallet_db.py:321]: it seems a bit odd to put test functions in here

Comment (aqk) [reply chia/cmds/check_wallet_db.py:321]: Yes - Just pushing this up so people can use it. Still need to move some things around, and add some tests.

**Comment by trepca** [comment]
@aqk how about `chia wallet doctor` :)

**Comment by aqk** [comment]
> @aqk how about `chia wallet doctor` :)

:)

File: chia/cmds/check_wallet_db.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+# TODO: Add verbose: print start and end derivation index per wallet
+# TODO: Check for missing paired wallets (eg. No DID wallet for an NFT)
+# TODO: Print name and ID of WalletType
+# TODO: Use fancy command line library
+# TODO: Check used contiguous
+# TODO: Check for missing DID Wallets
+# TODO: Use require_derivation_paths() to see whcih Wallets need derivations
+
+
+def check_for_gaps(array: List[int], start: int, end: int, *, data_type: str = "Element") -> List[str]:
```
Comment (aqk) [inline chia/cmds/check_wallet_db.py:25]: Should I make this generic and put it in `chia.util` ?

File: chia/cmds/check_wallet_db.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+            print("\nWallet DB Tables:")
+            print(*([r[0] for r in await cursor.fetchall()]), sep=",\n")
+            print("\nWallet Schema:")
+            print(*(await (await cursor.execute("PRAGMA table_info('users_wallets')")).fetchall()), sep=",\n")
+            print("\nDerivationPath Schema:")
+            print(*(await (await cursor.execute("PRAGMA table_info('derivation_paths')")).fetchall()), sep=",\n")
+            print()
+            return []
+
+    async def check_wallets(self) -> List[str]:
```
Comment (aqk) [inline chia/cmds/check_wallet_db.py:139]: `check_wallets` could be better - I  could finish moving all the SQL calls up front.

**Comment by wallentx** [comment]
This is excellent!
What was the thinking behind placing this as an option under the `chia wallet` command, rather than `chia db`?

**Comment by aqk** [comment]
> This is excellent! What was the thinking behind placing this as an option under the `chia wallet` command, rather than `chia db`?

Ah ha, I didn't know about `chia db`. Looks like that currently means "blockchain db", so we'd need a way to say which db we were talking about.

As for what to do if you have a problem: well, you already have a problem, but now you have more info. So, same support avenues as we have now. When we start to understand *how* the Wallet DB is being damaged, we should probably add an option to this command to back up and fix the db.

**Review by wjblanke — APPROVED** [review]
aok

**Pull request merged**